### ISSUE TITLE
Errors with generated go_repositories could not be resolved

### ISSUE THREAD

**Issue by StyleASD** [bazelbuild/bazel-gazelle#525]
I found an issue while trying to import the following lib with go modules `github.com/terraform-providers/terraform-provider-azurerm`. 

The issue seems to come from a dependency this repository has on `github.com/Azure/go-autorest` 

I have created an example repo to help reproduce [here](https://github.com/StyleASD/example_repo) 

Error Output:
```
ERROR: /.cache/bazel/_bazel_pvt/e3eb6d659947ddbf18a1f3aa6abeed98/external/com_github_terraform_providers_terraform_provider_azurerm/azurerm/BUILD.bazel:3:1: no such package '@com_github_azure_go_autorest_autorest//': The repository '@com_github_azure_go_autorest_autorest' could not be resolved and referenced by '@com_github_terraform_providers_terraform_provider_azurerm//azurerm:go_default_library'
ERROR: Analysis of target '//:go_default_library' failed; build aborted: no such package '@com_github_azure_go_autorest_autorest//': The repository '@com_github_azure_go_autorest_autorest' could not be resolved

```
I think the issue is that each subfolder in `github.com/Azure/go-autorest` has its own go.mod and `@github.com/terraform-providers/terraform-provider-azurerm` seems to generate the BUILD.bazel files with `@com_github_azure_go_autorest_autorest//:go_default_library` instead of `@com_github_azure_go_autorest//autorest:go_default_library`

**Comment by jayconrod** [comment]
This is a complicated problem. When `go_repository` runs Gazelle to generate build files for `com_github_hashicorp_terraform`, it attempts to resolve external import paths to labels without knowledge of what modules are being used by the main workspace. When it encounters the path `github.com/Azure/go-autorest/autorest`, it runs `go list -find -f {{.Module.Path}} -- github.com/Azure/go-autorest/autorest` in an empty module. This package is provided by both `github.com/Azure/go-autorest/autorest` and `github.com/Azure/go-autorest` at various versions. The first one is longer, so that wins. Gazelle converts the module path to the repository name `com_github_azure_go_autorest_autorest`. However, you're not actually using module `github.com/Azure/go-autorest/autorest`, so Bazel reports an error when it sees a label mentioning this repository.

I've opened #529 to address the root cause of this issue. When Gazelle runs inside `go_repository`, it should configure itself using the main WORKSPACE. If it did this, it would know that `github.com/Azure/go-autorest` is the longest prefix for this import path, and that corresponds to `com_github_azure_go_autorest`.

As a workaround, you can set `build_extra_args = ["-known_import=github.com/Azure/go-autorest"]` on the `go_repository` rule for `com_github_hashicorp_terraform`. Unfortunately, that only works for that specific module, within that specific `go_repository`. Many more of these flags may be needed.

**Comment by StyleASD** [comment]
Thank you for the detailed explanation. I will try and configure it using the workaround you have suggested.

**Comment by mariusgrigoriu** [comment]
With #529 closed, is a workaround still required?

**Comment by jayconrod** [comment]
Verified that this can be analyzed with the latest rules_go and gazelle, after re-importing the go.mod file.

It still can't be built because `github.com/terraform-providers/terraform-provider-azurerm` is a main package and has private visibility. That seems fine though.

**Comment by shimmerjs** [comment]
I still appear to hit this problem with decent regularity each time Kubernetes is released, e.g., today `bazel build //...` fails because it is trying to pull the latest version of kubernetes/client-go (despite me being pegged to 0.17.4  and my go_repositories not changing in about a week).  Am I misunderstanding how #529 is supposed to help and the expectation is we should be maintaining `known_import` args for our `go_repository`'s as issues arise?

**Comment by jayconrod** [comment]
@shimmerjs You should have a `go_repository` for every external repository you transitively depend on either in `WORKSPACE` or in a macro referenced by a directive in `WORKSPACE` like `# gazelle:repository_macro deps.bzl%go_deps`. You can also use `# gazelle:repository` directives to declare a repository that's not fetched with `go_repository` or is in a place that Gazelle can't understand. When Gazelle is invoked in `go_repository`, it uses that information to resolve dependencies.

If that's not working for you, please open a new issue.

**Comment by shimmerjs** [comment]
I am doing all of those things.  I would be happy to open a new issue, but since it was resolved by adding `"-known_import=k8s.io/api"` to the generated `go_repository` rule for `io_k8s_client_go`, I was under the impression it might be the same issue.

It appears that `@io_k8s_api//auditregistration/v1alpha1:go_default_library` isn't resolvable by Gazelle and that causes Gazelle to go out to the network and resolve the latest module version (per your comment https://github.com/bazelbuild/bazel-gazelle/issues/545#issuecomment-501716695, although I may be misunderstanding).  I also tried a `gazelle:resolve` directory to map `k8s.io/api/auditregistration/v1alpha1` to `@io_k8s_api//auditregistration/v1alpha1:go_default_library`, but that didn't resolve the issue.  Only the `known_import` trick worked.

FWIW, the label does exist:

```sh
λ bazel query @io_k8s_api//...
@io_k8s_api//flowcontrol/v1alpha1:go_default_library
@io_k8s_api//discovery/v1beta1:go_default_library
@io_k8s_api//discovery/v1alpha1:go_default_library
@io_k8s_api//core/v1:go_default_test
@io_k8s_api//auditregistration/v1alpha1:go_default_library
...
```

**Comment by mariusgrigoriu** [comment]
We also got hit with this error, which was caused by the same release. 

We're building an external repository that depends on the k8s client_go, which depends on `k8s.io/api`. All external repositories involved, including client_go and api are imported using `go_repository` rules in a .bzl file loaded in `WORKSPACE`, but not directly in the top-level `WORKSPACE` file.

It seems that the fix from #529 literally reads only the `WORKSPACE` file and no .bzl files loaded by it. We resolved the above issue by moving the `go_repository` rule related to `io_k8s_api` into the `WORKSPACE` file.

Would it be a valid feature request to extend #529 to also read any repos loaded through other .bzl files?

**Comment by jayconrod** [comment]
Please file issues for these questions. I can't really help you if I can't see the repositories or WORKSPACE files in question.

> It seems that the fix from #529 literally reads only the WORKSPACE file and no .bzl files loaded by it. We resolved the above issue by moving the go_repository rule related to io_k8s_api into the WORKSPACE file.

It also reads files mentioned with `# gazelle:repository_macro`. Use that if your dependencies are in another file. The file must be in the same repository though.

> Would it be a valid feature request to extend #529 to also read any repos loaded through other .bzl files?

Probably not. In general, Gazelle won't evaluate Starlark code or read .bzl files from other repositories, so it's limited in what it can do. Having an explicit `# gazelle:repository_macro` declaration seems better to me.
